### ISSUE TITLE
New Developer Guide + Code map

### ISSUE THREAD

**Issue by dnjohnstone** [hyperspy/hyperspy#782]
There is an inevitable learning barrier to overcome for those wanting to add functionality to HyperSpy. It seems to me that this can be quite daunting and I know some Electron Microscopists who have thought of contributing to HyperSpy but have been put off entirely by this barrier...

For many microscopists (including myself) a large part of this barrier is not knowing what is required to contribute to a large open-source project -- by which I mean things like, what a test is and how to write one, what documentation is required, even how to use git. Obviously much of this information is available somewhere but I don't think it is clear to someone not already well versed in Python where to look for this..

Specifically to HyperSpy there are some rather convoluted bits of code structure that are tricky to work out starting from scratch and perhaps some 'code map' might be very helpful.

A suggestion to combat this is a blog or blog aggregate associated with HyperSpy that regular contributors can use to collectively help the community understand the code.

What do people think? Is it just me who found this hard/time-consuming?
## TODO
- [x] Developer Guide
- [ ] Code map

**Comment by francisco-dlp** [comment]
You're totally right. A good developer guide is very much needed. So much that I'll categorize this as a bug and add it to the 0.9 milestone.

**Comment by vidartf** [comment]
I agree, but think your original proposal is actually two-fold:
 1. Developer guide.
 2. Code map.

I'd like to add some suggestions for the code map: In principle, this would be very similar (identical?) to a quick-start guide for users. It should outline quickly the architecture of Hyperspy, and where you should expect to find different bits of functionality (and why!). This would help users find functionality, and contributors to figure out where new code would fit best. I'd suggest to put this as the second page in the users guide (after introduction).

**Comment by francisco-dlp** [comment]
+1000 on that. I'll edit the header with a TODO list.

**Comment by dnjohnstone** [comment]
Ok 

Developer guide --> I'm happy to put together an initial draft answering questions that I have found and do find myself having. Where do you want me to put that though?

Code map --> this probably needs input from someone who knows the code better than I do although happy to help if I can.

**Comment by francisco-dlp** [comment]
Great. `hyperspy/doc/developer_guide` seems to me the natural place to place the DG.

**Comment by francisco-dlp** [comment]
One thing that we should agree about and write in the dev guide: how do we promote contributors to members of the dev team? Also, which are the benefits of being a member of the dev team.

This is what I propose to start the discussion:

The most important role of a member of the dev team is actually reviewing and merging. Hence, I propose to invite someone to join the dev team when:
* Has contributed some PRs and shows good understanding of HyperSpy internals
* Has reviewed/helped review some PRs.

Also we should set an idle time (np PR, and, more importantly no review) span for members of the dev team, after which they'll be invited to join the Emeritus team. Maybe 6 months/1 year...?

You get it: the aim is to encourage contributors to review. 

Dev team membership benefits:

* Writing rights
* Access to the project's Google Analytics
* (anything that I'm missing, other than the honour and prestige ;) ?)

Also, as a little thank you from the hspy community to the top contributor of each release, I thought that we could give this developer the right to name to the release, maybe starting from 0.9. (Of course with veto rights for the dev team just in case...) What do you think?

**Comment by dnjohnstone** [comment]
Ok - I'll make a start on the developer guide then with the intention of having a decent pull request together by Christmas. But does that mean we're steering away from a blog / blog aggregate? In some ways I thought that was nice as it would allow people to develop personal(professional) websites where their individual contribution is clear.

Promotion to the dev team is probably a bit above my station but your suggestion sounds sensible to me - it seems clear that reviewing is a substantial bottleneck that's creating work repetition so your points in that regard seem particularly sensible to give some incentive to do it. That said perhaps the issue is simply number of people rather than idleness...

**Comment by dnjohnstone** [comment]
Also - just noticed that this exists: https://github.com/hyperspy/hyperspy/wiki/Guidelines-for-developers

Probably a sensible starting point

**Comment by thomasaarholt** [comment]
Sweet. Good reading.
Thomas Aarholt
thomasaarholt@gmail.com

**Comment by tjof2** [comment]
I also agree with this - knowing where/how to start would be amazing!

**Comment by dnjohnstone** [comment]
Thanks for feedback - given the interest in this I don't want to lose that momentum so I'll accelerate my plan and knock out a first draft over the weekend that can at least serve as a framework for conversation between those starting out...

So @francisco-dlp where do we want this?

/doc/dev_guide.py

OR wiki https://github.com/hyperspy/hyperspy/wiki/Guidelines-for-developers

OR blog (I can start a personal one trivially that could then be aggregated)

**Comment by dnjohnstone** [comment]
PS. picking up on the very useful comment from @vidartf his suggestion for location of the code map as the second page of the user-guide after the introduction makes a lot of sense to me as it will be useful to all users and potential contributors.

In a separate pull request I'll put an  outline for what that page might look like so that we can agree on structure before populating it with the real information

**Comment by francisco-dlp** [comment]
+1 on putting this in `doc/developers_guide.rst` We should get rid of the
wiki once this is ready.

Starting a blog aggregator for hspy sounds cool. That would be the ideal
container for recipes, tricks etc., but the dev guide certainly belongs to
the docs.

2015-12-11 10:10 GMT+00:00 dnjohnstone <notifications@github.com>:

**Comment by vidartf** [comment]
From #445:
 - [ ] Reporting bugs
 - [ ] PR checklist:
   - [ ] best practices
   - [ ] Where to PR: master and maintenance branches
 - [ ] Milestones: wishlists

**Comment by vidartf** [comment]
Not sure if this issue is finished or not, but either way it could be useful to copy/move the developer guide or parts of it to a CONTRIBUTING.md file. See https://github.com/blog/1184-contributing-guidelines.

**Comment by dnjohnstone** [comment]
I came across this recently - which I think is a particularly good example.... Much of the github stuff is exactly the same for HyperSpy too.

http://tomopy.readthedocs.io/en/latest/devguide.html
